### ISSUE TITLE
Audit of Features missing in JupyterBook

### ISSUE THREAD

**Issue by rowanc1** [executablebooks/mystjs#189]
Although mystjs and JupyterBook/sphinx are targeted at slightly different use-cases, there is overlap in creating websites (see [background](https://myst-tools.org/docs/mystjs/background#how-do-jupyter-book-and-mystjs-relate)). Below is a list of missing features/directives/roles that are not yet implemented in `mystjs`, we hope to have parity at the rendering level first, then move on to extensibility (see #181 as a major step).

We will try to keep this updated as we audit more jupyterbooks.

**Directives**

- [ ] example
- [ ] toggle (show option)
- [ ] literalinclude
- [ ] prf
- [ ] only
- [ ] panels
- [ ] epigraph (replicated mostly with quote improvements)
- [ ] glossary / terms
- [ ] contents
- [ ] doc
- [ ] field lists
- [ ] tasks #187 
- [x] girds (grid-item-card)
- [x] cards (header/footer)
- [x] dropdown
- [ ] sphinx proof

**Extensibility**

See #181 as a roadmap.

- [ ] transforms turned on (e.g. dollarmath, etc.)
- [ ] directives / roles

**Execution**

- [ ] Linking artifacts from a notebook (glue) - we are supporting this using ids/imports ([see docs](https://myst-tools.org/docs/mystjs/cross-references#targeting-cells))
- [ ] running code execution for a notebook

**Improvements**

- [ ] code-cell: tags (output_scroll, hide)
- [x] admonition: classes
- [x] admonition: dropdown class
- [x] intersphinx

**UI**
- [ ] TOC - collapse
- [ ] logo more like JB
- [ ] toc footer
- [ ] page footer
- [ ] banner
- [ ] search #100 
- [ ] internationalization #166 

URL:
* Collapsing things down to a `project/slug` isn't how JB works, but probably ok

Export
- [ ] static html #188

**Comment by Jan-David-Black** [comment]
@rowanc1 To implement a feature on this list we need to simultaneously consider LaTeX and the Web output, correct? How could we go about sidebar/margin content? Using \minipage or similar does not really float my boat. At the moment I would be in favor of adding the sidebar content to the appendix, or to add it below the current paragraph as a content box.
What do you think?

**Comment by rowanc1** [comment]
It would certainly be appreciated to consider both html, latex and other exports, however, we can take contributions that get us there in a few iterations. The default falls back to rendering the child content, so it is just inline in the document. I think this might depend on what template you use and as you suggest, there isn't really a good default in latex!

If you are interested in implementing one of these features, can you open a specific issue on it? We can help give some guidance on where to look and how to get started!!

**Comment by Jan-David-Black** [comment]
Maybe I'll find some time to work on the sidebar soon. Trying to come up with a workaround I noticed that html tags are not left unchanged:
![grafik](https://github.com/executablebooks/mystmd/assets/92856893/8b2e1c4a-74db-4254-b068-6346bf94d8c6)

The myst-spec doesn't seem to say anything on how raw html is to be treated. The [cheatsheet](https://jupyterbook.org/en/stable/reference/cheatsheet.html) (which seems to be jb-flavoured myst) seems to indicate that html is interpreted as is.

**Comment by rowanc1** [comment]
I have opened and linked #516 with a few pointers to start, just ping me if you start working on it and get stuck and I can try to help out fast!

There is an outstanding bug for the HTML here:
- #418

It works fine when the round-trip is HTML (e.g. using the underlying js libraries), but our renderer is react, so we actually need the AST of the HTML correct, which is currently incorrect and not nested properly. For example, this [AST](https://mystmd.org/sandbox?tab=ast&myst=PABiAD4AYgBvAGwAZAAgAHQAZQB4AHQAPAAvAGIAPgA%3D) of bold is `[html, text, html]` rather than `[html > text]` or converted to `[strong > text]` in this simple case. This is still going to be a bit of work to sort through.

**Comment by Jan-David-Black** [comment]
Even with the "wrong" AST[^1] one seems to produce correct html with the `allowDangerousHtml` option:

```js
import { u } from 'unist-builder';
import { mystToHtml } from 'myst-to-html';

let ast = u('root', [u('html', '<b>'), u('text', 'test'), u('html', '</b>')]);

const skewed_html = mystToHtml(ast);
console.log(skewed_html); // test

const html = mystToHtml(ast, {'hast': {'allowDangerousHtml': true}, 'stringifyHtml': {'allowDangerousHtml': true}});
console.log(html); // <b>test</b>
```

[^1]: The correct handling of html doesn't seem to be firmly defined: Does it all go into one html node, or using children for the content. How is markdown syntax inside html handeled etc.

**Comment by rowanc1** [comment]
Yes, it will work for explicit HTML outputs which is string concatenation! However most of the tools are built on React, which operates on the AST and will try to create `<b>` as HTML, then `'test'` text, then `</b>`. Which will end up with `<b></b>test` as the first node closes by default and the closing node isn't rendered. I think it is a bit worse actually as it looks like there is an extra paragraph, `<b>test</b>` turns into:

![image](https://github.com/executablebooks/mystmd/assets/913249/9dd9b667-9571-469c-97d4-54593f9ade77)

Not great, and something to fix up soon (#418)!

**Comment by holmrenser** [comment]
Not sure this is the right place to report, but here goes:
In Jupyterbook1 images can be clicked on, which redirects to a static page of the full image. This is especially useful for zooming in on high-res figures. Jupyterbook2 does not (yet) do this.

**Comment by rowanc1** [comment]
Thanks @holmrenser!

**Comment by parmentelat** [comment]
hiya lads

I have JB1 documents that come with static assets - typically .mp4 videos, or static HTML documents like folium maps, ...
under JB1 I could simply store them under `_static` and I could then find those at a URL like 
BASE_URL/_static/foo.html 

as far as I can tell this is no longer the case under jb2
how can I mimick a similar behaviour with jb2, i.e. define that some specific folder essentially needs to ship as-is in the build ?

**Comment by choldgraf** [comment]
@parmentelat good point! I've created a parent issue to track static links etc in document builds here, and added it to the list above:

- https://github.com/jupyter-book/mystmd/issues/1921

**Comment by parmentelat** [comment]
thanks @choldgraf 
while we're at it may I try to ping the team on 
https://github.com/jupyter-book/mystmd/issues/1164
which I understand is no hot thing of course, but would be quite handy :)

**Comment by parmentelat** [comment]
on the same track I notice that url-defined items in the toc seem to not yet be implemented
I mean:

```yaml
    - title: chapter
      children:
        - file: plain-file.md
          title: a title set in the toc
        - url: https://some-other-page/on-the-web
          title: link to external resource
```

which is something that was achievable with jb1

(not sure if that helps or not, but in my use case this is meant to open in a separate tab
 iirc I had to tweak jb1 to achieve that behaviour, not 100% sure though)

**Comment by parmentelat** [comment]
as an aside, let me maybe give some background
I had tried jb2 a while back, like a year ago, but had not been able to adopt it for the current academic year
I'm giving it another try and am feeling rather confident this time
I am using this thread to summarize the pebbles left in my shoe
let me know if I should use another medium for this

**Comment by itcarroll** [comment]
I'm attempting to update a jb1 to a jb2, and snagged on different handling of code-cell tags. Notably [jb1](https://jupyterbook.org/en/stable/reference/cheatsheet.html#tags) recognizes "scroll-output" and [jb2](https://mystmd.org/guide/notebook-configuration#tbl-notebook-cell-tags) does not. I do wonder though, if "scroll-output" is missing because a more correct reliance on the notebook [cell metadata](https://nbformat.readthedocs.io/en/latest/format_description.html#cell-metadata) keyword "scrolled" is planned? Somehow, I would like to keep scrolling!

**Comment by Alexboiboi** [comment]
Hi there,

The code-cell tag `full-width` works in jb1 but not in jb2, does this fit in this discussion?

**Comment by scottyhq** [comment]
> I'm attempting to update a jb1 to a jb2, and snagged on different handling of code-cell tags. Notably [jb1](https://jupyterbook.org/en/stable/reference/cheatsheet.html#tags) recognizes "scroll-output" and [jb2](https://mystmd.org/guide/notebook-configuration#tbl-notebook-cell-tags) does not. I do wonder though, if "scroll-output" is missing because a more correct reliance on the notebook [cell metadata](https://nbformat.readthedocs.io/en/latest/format_description.html#cell-metadata) keyword "scrolled" is planned? Somehow, I would like to keep scrolling!

Just wanted to second this, as text outputs can get quite long!

**Comment by Klemet** [comment]
Hello folks ! I apologize if I've missed some information, but I've spent hours trying to understand how to transition to jb2 properly and haven't been succeeding so far. Wanted to know if these features are missing, or if I'm missing the instructions for them.

I've got a jb1 notebook (html static [here](https://klemet.github.io/Calibration_PnET_DIVERSE/intro.html)). Was super enthusiastic about going to jb2, which looked amazing. I followed the instructions to install Jupyter Book v2.0.0-b1 (which seems to be the latest at current date).

I then tried to follow the instructions [here](https://next.jupyterbook.org/upgrade) to upgrade my book. It seems to have worked, but a lot of my settings has been lost in the transition.

In particular, my use case was to build the HTML static website into a /doc folder. I did that before with the following in `_config.yml` : 

```yaml
repository:
  url: https://github.com/Klemet/Calibration_PnET_DIVERSE  # Online location of your book
  path_to_book: docs  # Optional path to your book, relative to the repository root
  branch: master  # Which branch of the repository should be used when creating links (optional)
```

With jb2, I'm unable to generate a static HTML website or to output it where I want. I've been reading the Myst documentation to try and find how to write things down in `myst.yml`; but to no avail. Not only am I not finding the right command to simply output the HTML files somewhere (I can use `jupyter book build --site`, but I'm not finding the HTML files it's generating); and I can't find the right parameters for `myst.yml` to indicate the output folder.

If anybody has a pointer, I'd be very grateful !

**Comment by rowanc1** [comment]
Hi @Klemet - thanks for the comment. You are probably looking for `jupyter book build --html`, there isn't currently a way to specify the build location of the html. However you can copy it out later if you would like it in a different place when  (it is in `_build/html`); you can use a terminal command to move the folder after it is built. You will also need to set the BASE_URL appropriately. If you have issues on that (not seeing the CSS work) can you open a discussion on that specifically, happy to help!

**Comment by Klemet** [comment]
Thanks a lot, @rowanc1 😁 !

I was able to make something with the following code cell in Jupyter lab : 

```shell
%cd /calibrationFolder/calibrationJupyterBook_JupyterBookv2

%env BASE_URL=https://klemet.github.io/Calibration_PnET_DIVERSE

!ls

!jupyter book build --html

!rm -rf ../docs/*

!mv ./_build/html/* ../docs
```

It's updated on Github pages, and seems to display well. However, the navigation links are non-fonctional because of a caps issue.

For example, the navigation link to access the first page of my documentation (as generated by Jupyter book v2) is :

```
https://klemet.github.io/Calibration_PnET_DIVERSE/startingcalibrationtest/intro-calibrationtest
```
But it's not working, and Github pages returns a 404 error.

The right link should be :

```
https://klemet.github.io/Calibration_PnET_DIVERSE/StartingCalibrationTest/intro-calibrationtest
```
with `StartingCalibrationTest` having the right caps.

The links in the TOC on my first page made with : 

```md
```{tableofcontents}
``` 

 also have a point (`.`) instead of a `/` for the last part of the path. They look like :

```
https://klemet.github.io/Calibration_PnET_DIVERSE/startingcalibrationtest.loading-functions
```

I don't know if I should make a new issue out of this one ? Or if I'm again missing something ?

EDIT : Nevermind the issue about case/caps, : Jupyter book changed the folder name correctly, but Git didn't because of https://github.com/orgs/community/discussions/19561 . Should be able to fix this !

**Comment by Klemet** [comment]
I fixed the issue with the links, which was indeed due to Git not changing the folder name case.

Now, I see several issues remaining so far with jb2 on my side : 

- TOC made with `{tableofcontents}` has a point instead of a `/` for the last part of the URL, breaking the links.
- Long cell outputs cannot be made scrollable with `scroll-output` anymore as [itcarroll](https://github.com/itcarroll) mentionned.
- Large markdown tables are not scrollable horizontally anymore and are displayed in a wrong way : 

<img width="1648" height="872" alt="Image" src="https://github.com/user-attachments/assets/06c3c1ad-d335-46c7-b523-12ce13e88ed2" />

- Footnotes that are referenced at several places in a given document have a lot of "return" icon/button at the bottom of the page : 

<img width="1511" height="853" alt="Image" src="https://github.com/user-attachments/assets/4930c3fe-ac00-4733-90be-40012a926c5a" />

- Minor : but the navigation menu on the left doesn't "open up" a section when visiting the pages of a section, which will surely be confusing for some users.

Feel free to tell me if I should open issues properly for some of these. In any case, thanks for the help !

**Comment by stevejpurves** [comment]
> I'm attempting to update a jb1 to a jb2, and snagged on different handling of code-cell tags. Notably [jb1](https://jupyterbook.org/en/stable/reference/cheatsheet.html#tags) recognizes "scroll-output" and [jb2](https://mystmd.org/guide/notebook-configuration#tbl-notebook-cell-tags) does not. I do wonder though, if "scroll-output" is missing because a more correct reliance on the notebook [cell metadata](https://nbformat.readthedocs.io/en/latest/format_description.html#cell-metadata) keyword "scrolled" is planned? Somehow, I would like to keep scrolling!

It would be great to get an issue opened on juptyer-book/myst-theme for this, either to support `scroll-output` or the `scrolled` tag/metadata

**Comment by mforbes** [comment]
- [ ] [`tab-set-code`](https://sphinx-design.readthedocs.io/en/latest/tabs.html#tabbed-code-examples) blocks from sphinx-design are a really nice feature, especially since they synchronize across cells.  I.e., if you select one language, then all code examples will be displayed in that language throughout the page.  This does not seem to happen for regular tab sets.

Additionally, it would be awesome if a single notebook could execute code in multiple languages so these `tab-set-code` examples could be actually demonstrated, but this is a discussion for later.

**Comment by mforbes** [comment]
- [ ] The [`margin` tag](https://jupyterbook.org/en/stable/content/layout.html#margins-with-code-cells) is ignored in code cells:

````markdown
```{code-cell}
:tags: [margin]
print("This should appear in the margin")
```
````

Note, however, that -- unlike JupterBook -- you can nest code cells in other directives with MyST-MD, thus a workaround is

````markdown
:::::{margin}
```{code-cell}
print("This should appear in the margin")
```
:::::
````

**Comment by scottyhq** [comment]
Adding a :target: or :link: parameter for the {image} directive is possible in jupyterbook1 but not mystmd (see https://github.com/jupyter-book/mystmd/issues/1695#issuecomment-2529120352)
````
```{image} foo.png
:target: https://google.com OR `myst-ref`
```
````

**Comment by JimMadge** [comment]
Adding a note here that I wanted to use `guilabel` and found it missing. I won't open a new issue, because it is tracked here and I'm not planning to implement it.

@choldgraf and @fperez discussed this in Discord. The feeling was when we come to implement this, we would like to change the recommended name of the role to something more clear and generic, perhaps `uilabel` or just `ui`. As we use these elements to represent all kinds of ui interaction, not just gui. I imagine we would leave `guilabel` as an alias for compatibility with Sphinx.

**Comment by mforbes** [comment]
In the table, intersphinx is listed as complete, but I still don't see how to use standard roles etc.

* <code>{func}\`numpy.genfromtxt\`</code>: Fails with `unknown role: func`.  This is what I use extensively in my JupyterBook documentation and so would like to work in JupyterBook2.
* `<xref:numpy#numpy.genfromtxt>`: Works, but only displays [numpy](https://numpy.org/doc/stable/reference/generated/numpy.genfromtxt.html#numpy.genfromtxt).  There is no relevant title.
* <code>[\`numpy.genfromtxt\`](xref:numpy#numpy.genfromtxt)</code>: Works, but is almost worth not bothering with intersphinx and just doing it manually.

Am I missing something?

**Comment by mforbes** [comment]
* [ ] `execute: exclude_patterns:` Provide a way to exclude files from execution without excluding them from the documentation. https://github.com/orgs/jupyter-book/discussions/2475
